packages/flutter/lib/src/widgets/slotted_render_object_widget.dart

```diff
@@ -231,14 +240,47 @@ class SlottedRenderObjectElement<S> extends RenderObjectElement {
     }(), '${widget.runtimeType}.slots must not change.');
     assert(slottedMultiChildRenderObjectWidgetMixin.slots.toSet().length == slottedMultiChildRenderObjectWidgetMixin.slots.length, 'slots must be unique');
 
+    final Map<Key, S> oldKeyedSlots = _keyToSlot;
+    _keyToSlot = <Key, S>{};
+
+    // New slots for children that are moved to a different slot by key
+    // reparenting. It shouldn't be updated in place because two children could
+    // swap slots.
+    final Map<S, Element> slotToKeyedChild = <S, Element>{};
+
+    // Process children that need to be moved to new slots because of key
+    // reparenting first.
     for (final S slot in slottedMultiChildRenderObjectWidgetMixin.slots) {
-      _updateChild(slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot), slot);
+      final Widget? widget = slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot);
```

Comment:
I dislike that we call `childForSlot` twice for unkeyed widgets. While the API doesn't make any promises how often this gets called, it my be expensive to construct the widget for a slot, so we should avoid this duplication...

Reply:
> as long as we don't match existing keyed widget with new unkeyed ones

There's nothing that prevents this I think?

Reply:
Also I think the type parameter should extend `Object`?

Reply:
> Also I think the type parameter S should extend Object?

Why?

Reply:
Regarding the one-pass, I was thinking we could do something like this (pseudo-code):

```dart
for (final S slot in slottedMultiChildRenderObjectWidgetMixin.slots) {
  final Widget? widget = slottedMultiChildRenderObjectWidgetMixin.childForSlot(slot);
  if (widget.key != null) {
    var oldChild = oldKeyToElement.remove(widget.key);
    if (oldChild == null) {
      oldChild = _slotToChild[slot].widget.key != null : _slotToChild[slot] : null;
    } else if (oldChild == _slotToChild[oldChild.slot]) {
       _slotToChild.remove(oldChild.slot);
    }
    var newChild = updateChild(oldChild, widget, slot);
    _slotToChild[slot] = newChild;
  } else {
   var oldChild = _slotToChild[slot].widget.key == null ? _slotToChild[slot] : null;
   var newChild = updateChild(oldChild, widget, slot);
   _slotToChild[slot] = newChild;
  }
}
// TODO: match the remaining children in oldKeyToElement to null with `updateChild`.
```

I am sure I messed up some details in that code, but something like that?

Reply:
> Why?

With that `Key` to `S` map I'm making the assumption that `S` cannot be null (https://master-api.flutter.dev/flutter/dart-core/Map/operator_get.html) so if people use `null` as a slot it could mess up the key to slot lookup.

> one-pass

Ahh I see if `oldWidget.key != null` and `newWidget.key == null` `Widget.canUpdate` will return false and we'll inflate a new element anyways. But if that happens and the old keyed `Elemenent` doesn't end up getting mapped to a new widget we'll have to call `deactivate` on it manually, so we kinda need a new collection to do the bookkeeping anyway?

Reply:
> so if people use null as a slot it could mess up the key to slot lookup.

Good point. In theory, null is a valid slot value, which I think the current implementation fully supports. However, if changing this isn't breaking anything (and it really simplifies the implementation a lot - which I am not fully convinced yet) I would be ok with changing this and documenting that limitation.

> But if that happens and the old keyed Element doesn't end up getting updated by a new widget we'll have to call deactivate on it manually, so we kinda need a new collection to do the bookkeeping anyway?

Not sure I understand. If we pass an Element to `updateChild` it will either get re-used or properly deactivated.

Oh, but I think I missed a case in my code above (maybe that's the one you meant): When a keyed child moves into a slot previously occupied by an unkeyed child, we need to deactivate that child by calling `updateChild(unkeyedChild, null, slot)` for it before updating the slot of the keyed child. But I think we can just do that right before the first `var newChild = updateChild(oldChild, widget, slot);` occurence in the code above?

Reply:
Updated. Also tried to avoid using `S` as map value type.